|                |   |
:----------------|---|
| **Nombre**     | Valeria Guzmán Galván
| **Fecha**      | 19/02/26 |
| **Expediente** | 756902 |

# T05 - Motor Trend Car Road Tests

Utiliza el archivo "Motor Trend Car Road Tests.xlsx" y completa las siguientes actividades:

In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from google.colab import files
uploaded = files.upload()


df = pd.read_excel("Motor Trend Car Road Tests.xlsx")
df.head()

Saving Motor Trend Car Road Tests.xlsx to Motor Trend Car Road Tests (2).xlsx


model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

### 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

* Calcula el R2 e interpreta los signos de los betas.
* Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.
* Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [ ]:
df = df.drop(columns=["model"])

In [ ]:
X = df.drop(columns=["mpg"])
y = df["mpg"]

In [ ]:
lr1 = LinearRegression()
lr1.fit(X, y)

y_pred = lr1.predict(X)

r2_1 = r2_score(y, y_pred)
print("R2:", r2_1)

R2: 0.8690157644777647


Esto significa que el modelo explica aproximadamente 86.9% de la variabilidad de mpg

In [ ]:
print("Intercepto:", lr1.intercept_)
print("Betas:")
for i in range(len(X.columns)):
    print(X.columns[i], ":", lr1.coef_[i])

Intercepto: 12.30337415599622
Betas:
cyl : -0.11144047788686436
disp : 0.013335239913341172
hp : -0.02148211898913649
drat : 0.7871109722361131
wt : -3.7153039283274794
qsec : 0.8210407496746295
vs : 0.31776281418541313
am : 2.5202268872084304
gear : 0.6554130170817877
carb : -0.1994192548562635


**Interpretación de los betas**

Beta positivo → cuando la variable aumenta, mpg aumenta

Beta negativo → cuando la variable aumenta, mpg disminuye

**Train-test-split donde se use el 40% de los datos para entrenar**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.40, random_state=42)

lr_split = LinearRegression()
lr_split.fit(X_train, y_train)
r2_train = lr_split.score(X_train, y_train)
r2_test = lr_split.score(X_test, y_test)

print("R2 entrenamiento:", r2_train)
print("R2 prueba:", r2_test)

R2 entrenamiento: 0.9982122857062323
R2 prueba: -7.1071406535106725


El modelo tiene un R2 muy alto en entrenamiento, pero muy bajo en prueba. Esto significa que el modelo aprendió demasiado bien los datos de entrenamiento, pero no funciona bien con datos nuevos. Esto es un sobreajuste.

**Regularización L2**

In [ ]:
ridge_01 = Ridge(alpha=0.1)
ridge_01.fit(X_train, y_train)

print("Lambda = 0.1")
print("R2 train:", ridge_01.score(X_train, y_train))
print("R2 test:", ridge_01.score(X_test, y_test))

Lambda = 0.1
R2 train: 0.9794200798509615
R2 test: 0.26067024754956425


In [ ]:
ridge_1 = Ridge(alpha=1)
ridge_1.fit(X_train, y_train)

print("Lambda = 1")
print("R2 train:", ridge_1.score(X_train, y_train))
print("R2 test:", ridge_1.score(X_test, y_test))

Lambda = 1
R2 train: 0.9278776695116554
R2 test: 0.6310793203589048


In [ ]:
ridge_10 = Ridge(alpha=10)
ridge_10.fit(X_train, y_train)

print("Lambda = 10")
print("R2 train:", ridge_10.score(X_train, y_train))
print("R2 test:", ridge_10.score(X_test, y_test))

Lambda = 10
R2 train: 0.8633880241271138
R2 test: 0.6567019278519771


Al agregar regularización L2, el R2 de prueba mejora considerablemente y deja de ser negativo. Esto indica que la regularización ayuda a reducir el sobreajuste. Conforme aumenta lambda, el modelo se vuelve más estable y generaliza mejor.

### 1.2 Repite el ejercicio anterior usando 'qsec' como salida.

In [ ]:
X_qsec = df.drop(columns=["qsec"])
y_qsec = df["qsec"]

In [ ]:
lr2 = LinearRegression()
lr2.fit(X_qsec, y_qsec)

y_pred2 = lr2.predict(X_qsec)

r2_2 = r2_score(y_qsec, y_pred2)
print("R2:", r2_2)

R2: 0.8746925777094


El modelo obtuvo un R2 de 0.875 aproximadamente, lo que significa que explica cerca del 87.5% de la variabilidad de qsec.

In [ ]:
print("Intercepto:", lr2.intercept_)
print("Betas:")
for i in range(len(X_qsec.columns)):
    print(X_qsec.columns[i], ":", lr2.coef_[i])

Intercepto: 17.77617692970135
Betas:
mpg : 0.06904768088540954
cyl : -0.36267823342467104
disp : -0.007500837721945819
hp : -0.0015625524378298783
drat : -0.1310635991489382
wt : 1.4963322381970525
vs : 0.9700347389620253
am : -0.9011864104662037
gear : -0.20128511286790152
carb : -0.27359751219248263


**Interpretación de los betas**

Beta positivo → cuando la variable aumenta, qsec aumenta
Beta negativo → cuando la variable aumenta, qsec disminuye

Las variables con coeficientes positivos, como mpg, wt y vs, indican que al aumentar estas variables, el tiempo en el cuarto de milla tiende a aumentar.
En cambio, variables con coeficientes negativos como cyl, disp, gear y carb indican que al aumentar estas variables, el tiempo qsec disminuye.

**Train-test-split donde se use el 40% de los datos para entrenar**

In [ ]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_qsec, y_qsec, train_size=0.40, random_state=42)

lr_split2 = LinearRegression()
lr_split2.fit(X_train2, y_train2)

r2_train2 = lr_split2.score(X_train2, y_train2)
r2_test2 = lr_split2.score(X_test2, y_test2)

print("R2 train:", r2_train2)
print("R2 test:", r2_test2)

R2 train: 0.9989478049029176
R2 test: -1.0013422169955328


El modelo tiene un R2 muy alto en entrenamiento, pero negativo en prueba. Esto indica que el modelo se sobreajusta y no generaliza bien cuando se evalúa con datos nuevos.

**Regularización L2**

In [ ]:
ridge2_01 = Ridge(alpha=0.1)
ridge2_01.fit(X_train2, y_train2)

print("Lambda = 0.1")
print("R2 train:", ridge2_01.score(X_train2, y_train2))
print("R2 test:", ridge2_01.score(X_test2, y_test2))

Lambda = 0.1
R2 train: 0.985578559916391
R2 test: 0.6878327440279519


In [ ]:
ridge2_1 = Ridge(alpha=1)
ridge2_1.fit(X_train2, y_train2)

print("Lambda = 1")
print("R2 train:", ridge2_1.score(X_train2, y_train2))
print("R2 test:", ridge2_1.score(X_test2, y_test2))

Lambda = 1
R2 train: 0.9347243870002221
R2 test: 0.7141258477214348


In [ ]:
ridge2_10 = Ridge(alpha=10)
ridge2_10.fit(X_train2, y_train2)

print("Lambda = 10")
print("R2 train:", ridge2_10.score(X_train2, y_train2))
print("R2 test:", ridge2_10.score(X_test2, y_test2))

Lambda = 10
R2 train: 0.8453882031592018
R2 test: 0.5044449278883241


Al agregar regularización L2, el modelo mejora considerablemente su capacidad de generalización. El R2 de prueba deja de ser negativo y alcanza su mejor valor con lambda igual 1. Esto indica que la regularización ayuda a reducir el sobreajuste y mejora el desempeño en datos nuevos.

### 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

* Calcula el R2 e interpreta los signos de los betas.
* Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [ ]:
df2 = df.copy()

In [ ]:
df2 = pd.get_dummies(df2, columns=["cyl", "gear", "carb"], drop_first=True)

In [ ]:
X2 = df2.drop(columns=["mpg"])
y2 = df2["mpg"]

In [ ]:
lr3 = LinearRegression()
lr3.fit(X2, y2)

y_pred3 = lr3.predict(X2)

r2_3 = r2_score(y2, y_pred3)
print("R2:", r2_3)

R2: 0.8930749320864843


El modelo con variables dummy obtuvo un R2 de aproximadamente 0.893, lo que indica que explica cerca del 89.3% de la variabilidad de mpg. Este valor es mayor que el modelo anterior sin dummies, por lo que la inclusión de variables categóricas mejora ligeramente el ajuste.

In [ ]:
print("Intercepto:", lr3.intercept_)
print("Betas:")
for i in range(len(X2.columns)):
    print(X2.columns[i], ":", lr3.coef_[i])

Intercepto: 23.87913244333949
Betas:
disp : 0.035546324039110955
hp : -0.0705068341268204
drat : 1.1828301846954223
wt : -4.529775844048
qsec : 0.36784482185575823
vs : 1.9308505366446092
am : 1.2121156988023376
cyl_6 : -2.64869527677977
cyl_8 : -0.3361629818023658
gear_4 : 1.1143549390497258
gear_5 : 2.5283959906632445
carb_2 : -0.9793543216488778
carb_3 : 2.999638749529034
carb_4 : 1.091422884400174
carb_6 : 4.4775692111484675
carb_8 : 7.25041125604525


**Train-test-split donde se use el 40% de los datos para entrenar**

In [ ]:
X_train3, X_test3, y_train3, y_test3 = train_test_split(X2, y2, train_size=0.40, random_state=42)

lr_split3 = LinearRegression()
lr_split3.fit(X_train3, y_train3)

r2_train3 = lr_split3.score(X_train3, y_train3)
r2_test3 = lr_split3.score(X_test3, y_test3)

print("R2 train:", r2_train3)
print("R2 test:", r2_test3)

R2 train: 1.0
R2 test: -1.3252877613458613


El modelo con dummies tiene un R2 perfecto en entrenamiento, pero negativo en prueba. Esto significa que el modelo se ajustó demasiado a los datos de entrenamiento y no funciona bien con datos nuevos.

### 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

In [ ]:
X3 = df2.drop(columns=["qsec"])
y3 = df2["qsec"]

In [ ]:
lr4 = LinearRegression()
lr4.fit(X3, y3)

y_pred4 = lr4.predict(X3)

r2_4 = r2_score(y3, y_pred4)
print("R2:", r2_4)

R2: 0.9082689440167556


El modelo con variables dummy obtuvo un R2 de aproximadamente 0.908, lo que indica que explica cerca del 90.8% de la variabilidad de qsec. Este valor es mayor que el modelo sin dummies, por lo que incluir variables categóricas mejora el ajuste.

In [ ]:
print("Intercepto:", lr4.intercept_)
print("Betas:")
for i in range(len(X3.columns)):
    print(X3.columns[i], ":", lr4.coef_[i])

Intercepto: 16.349137498224888
Betas:
mpg : 0.027741383917963425
disp : 0.003531156877540134
hp : -0.0020661884261466176
drat : 0.1078658136591686
wt : 0.8104721402955625
vs : 0.2657322079652916
am : -1.6942935129032517
cyl_6 : -1.1043425489180314
cyl_8 : -2.9669006817160866
gear_4 : 1.3322823397657453
gear_5 : 0.18231749209427014
carb_2 : -0.8344129784951013
carb_3 : -0.22930436729863302
carb_4 : -1.9470338112692946
carb_6 : -1.5652410384143138
carb_8 : -1.3323171730583514


**Train-test-split donde se use el 40% de los datos para entrenar**

In [ ]:
X_train4, X_test4, y_train4, y_test4 = train_test_split(X3, y3, train_size=0.40, random_state=42)

lr_split4 = LinearRegression()
lr_split4.fit(X_train4, y_train4)

r2_train4 = lr_split4.score(X_train4, y_train4)
r2_test4 = lr_split4.score(X_test4, y_test4)

print("R2 train:", r2_train4)
print("R2 test:", r2_test4)

R2 train: 1.0
R2 test: -0.05999228859335637


El modelo con dummies tiene un R² perfecto en entrenamiento, pero cercano a cero en prueba. Esto indica sobreajuste, aunque menos severo que en el caso anterior. El modelo no generaliza bien a datos nuevos.

### 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

En el ejercicio 1.1 el R2 fue 0.869 y en el 2.1 fue 0.893. Esto muestra que al agregar variables dummy el modelo explica mejor la variabilidad del mpg. Es decir, incluir las variables categóricas ayuda a mejorar el ajuste del modelo.

### 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

En el ejercicio 1.2 (sin dummies), el R2 fue aproximadamente 0.875, mientras que en el 2.2 (con dummies) fue aproximadamente 0.908.

Esto indica que al incluir variables dummy, el modelo logra explicar mejor la variabilidad de qsec. Es decir, las variables categóricas también aportan información útil para mejorar el ajuste del modelo.